### PR TITLE
Implementing Explicit Runge Kutta Methods 

### PR DESCRIPTION
**Pull request by tamaratambyah**
In this PR, I have implemented multi-stage, explicit Runge Kutta method in EXRungeKutta.jl. This follows the format and structure of RungeKutta.jl and IMEXRungeKutta.jl. 

I have added an additional operator `TransientEXRungeKuttaFEOperator`. However, I think this can be combined with `TransientRungeKuttaFEOperator` to have one operator which works for both explicit and diagonally implicit Runge Kutta (IMEX will probably still require its own operator). Once DIRK is working properly, I will look to tidy this up. 

Note, there are some preliminary implementations of DIRK on this branch. I have created a different branch for this now. 

@santiagobadia can you please review this PR. 

Thanks

### CODE REVIEW

File: my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl [hunk 1]
```diff
@@ -0,0 +1,97 @@
+using Pkg
```
Comment (santiagobadia) [inline my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl:1]: The whole my_tests folder should not be pushed to the main gridap project

File: my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
```
Comment (santiagobadia) [inline my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl:2]: Are all these tests related to the ExRK dev?
If not, we do not need these tests here.
I do recommend to start from a fresh Gridap fork, add the ExRK dev, and we start again

File: my_tests/dirk_step.jl [hunk 3]
```diff
@@ -0,0 +1,206 @@
+using Pkg
```
Comment (santiagobadia) [inline my_tests/dirk_step.jl:1]: Please try to stick to gridap code style guide, e.g., the way we name files here is not following it

File: src/Algebra/NLSolvers.jl [hunk 4]
```diff
@@ -52,7 +52,7 @@ function _solve_nr!(x,A,b,dx,ns,nls,op)
 
   # Newton-like iterations
   for nliter in 1:nls.max_nliters
-
```
Comment (santiagobadia) [inline src/Algebra/NLSolvers.jl:55]: Remove this (better start from scratch as indicated above)

File: src/ODEs/Exports.jl [hunk 5]
```diff
@@ -11,6 +11,8 @@ end
 @publish_gridapodes ODETools ThetaMethod
 @publish_gridapodes ODETools RungeKutta
 @publish_gridapodes ODETools IMEXRungeKutta
+@publish_gridapodes ODETools EXRungeKutta
+@publish_gridapodes ODETools DIRungeKutta
```
Comment (santiagobadia) [inline src/ODEs/Exports.jl:15]: better first EX -> accept
then we start with DI

File: src/ODEs/ODETools/DIRungeKutta.jl [hunk 6]
```diff
@@ -0,0 +1,140 @@
+"""
+Diagonally Implicit Runge-Kutta ODE solver
+** NOT WORKING YET **
```
Comment (santiagobadia) [inline src/ODEs/ODETools/DIRungeKutta.jl:3]: We cannot put in Gridap work in progress

File: src/ODEs/ODETools/ForwardEuler.jl [hunk 7]
```diff
@@ -53,6 +53,7 @@ function residual!(b::AbstractVector,op::ForwardEulerNonlinearOperator,x::Abstra
   vf = op.vf
   @. vf = (x-op.u0)/op.dt
   residual!(b,op.odeop,op.tf,(op.u0,vf),op.ode_cache)
+  println("residual: ", b)
```
Comment (santiagobadia) [inline src/ODEs/ODETools/ForwardEuler.jl:56]: remove

File: src/ODEs/TransientFETools/TransientFEOperators.jl [hunk 8]
```diff
@@ -436,10 +436,101 @@ function explicit_rhs!(
...
+  t::Real,
+  xh::T,
+  cache) where T
+  V = get_test(op)
+  v = get_fe_basis(V)
+  vecdata = collect_cell_vector(V,op.lhs(t,xh,v))
+  assemble_vector!(b,op.assem_t,vecdata)
+  b
+end
+
```
Comment (santiagobadia) [inline src/ODEs/TransientFETools/TransientFEOperators.jl:519]: remove below

File: src/ODEs/TransientFETools/TransientFEOperators.jl [hunk 9]
```diff
@@ -436,10 +436,101 @@ function explicit_rhs!(
...
+  test::FESpace
+  order::Integer
+end
+
+
+function TransientEXRungeKuttaFEOperator(lhs::Function,rhs::Function,jac::Function,
+  jac_t::Function,trial,test)
+  res(t,u,v) = lhs(t,u,v) - rhs(t,u,v)
+  assem_t = SparseMatrixAssembler(trial,test)
+  TransientEXRKFEOperatorFromWeakForm{Nonlinear}(res,lhs,rhs,(jac,jac_t),assem_t,(trial,∂t(trial)),test,1)
```
Comment (santiagobadia) [inline src/ODEs/TransientFETools/TransientFEOperators.jl:460]: Why nonlinear? I think that for a explicit method you just need a linear operator, that represents the mass matrix

File: src/ODEs/ODETools/RungeKutta.jl [hunk 10]
```diff
@@ -213,12 +214,24 @@ function zero_initial_guess(op::RungeKuttaNonlinearOperator)
   x0
 end
 
-function rhs!(op::RungeKuttaNonlinearOperator, x::AbstractVector)
-  u = x
```
Comment (santiagobadia) [inline src/ODEs/ODETools/RungeKutta.jl:217]: I see you are also trying to fix the standard RK method in this PR
It is ok, we can also include it here
But we must check that all tests pass

File: src/ODEs/ODETools/RungeKutta.jl [hunk 11]
```diff
@@ -213,12 +214,24 @@ function zero_initial_guess(op::RungeKuttaNonlinearOperator)
...
+function rhs!(op::RungeKuttaStageNonlinearOperator, x::AbstractVector)
+  u = zeros(eltype(x),length(x))
```
Comment (santiagobadia) [inline src/ODEs/ODETools/RungeKutta.jl:218]: We cannot do this, we are allocating a global vector here, it should go to the cache

File: src/ODEs/ODETools/RungeKutta.jl [hunk 12]
```diff
@@ -213,12 +214,24 @@ function zero_initial_guess(op::RungeKuttaNonlinearOperator)
...
+  for j in 1:op.i
+    @. u = u + op.a[op.i,j] * op.ui[j]
+  end
+  v = op.vi
+  @. v = (x-op.u0)/(op.dt)
+  rhs!(op.rhs,op.odeop,op.ti,(u,v),op.ode_cache)
+end
+
+function rhs!(op::RungeKuttaUpdateNonlinearOperator, x::AbstractVector)
+  u = zeros(eltype(x),length(x))
```
Comment (santiagobadia) [inline src/ODEs/ODETools/RungeKutta.jl:228]: same

File: src/ODEs/ODETools/RungeKutta.jl [hunk 13]
```diff
@@ -158,15 +159,14 @@ A(t,uf,∂uf/∂t) = ∂uf/∂t - ∑_{i<=s} b_i * f(ui,ti) = 0
 Uses the vector b as auxiliar variable to store the residual of the update ODE
 operator (e.g. identity or mass matrix), then adds the corresponding contribution from earlier stages.
 ```math
-b = [∂u/∂t]
-b - ∑_{i<=s} bi * f(ui,ti) = 0
+b = M(uf,tf)[∂u/∂t]
```
Comment (santiagobadia) [inline src/ODEs/ODETools/RungeKutta.jl:162]: What is uf here?
I don't understand this description of the problem
It does not look like RK

File: src/ODEs/ODETools/EXRungeKutta.jl [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+  d = solver.tableau.d
+
+  # Create cache if not there
+  if cache === nothing
+    ode_cache = allocate_cache(op)
+    vi = similar(u0)
+    ki = [similar(u0) for i in 1:s]
+    M = allocate_jacobian(op,t0,uf,ode_cache)
+    get_mass_matrix!(M,op,t0,uf,ode_cache)
+    nl_cache = nothing
```
Comment (santiagobadia) [inline src/ODEs/ODETools/EXRungeKutta.jl:39]: no need for this cache, we are storing nothing

File: src/ODEs/ODETools/EXRungeKutta.jl [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+  else
+    ode_cache, vi, ki, M, nl_cache = cache
+  end
+
+  nlop = EXRungeKuttaStageNonlinearOperator(op,t0,dt,u0,ode_cache,vi,ki,0,a,M)
```
Comment (santiagobadia) [inline src/ODEs/ODETools/EXRungeKutta.jl:44]: no nonlinear op here

File: src/ODEs/ODETools/EXRungeKutta.jl [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+  for i in 1:s
+
+    # solve at stage i
+    ti = t0 + c[i]*dt
+    ode_cache = update_cache!(ode_cache,op,ti)
+    update!(nlop,ti,ki[i],i)
+    nl_cache = solve!(uf,solver.nls,nlop,nl_cache)
```
Comment (santiagobadia) [inline src/ODEs/ODETools/EXRungeKutta.jl:52]: this is the LinearSolver cache, no nonlinear

File: src/ODEs/ODETools/EXRungeKutta.jl [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+"""
+ODE:    A(t,u,∂u  = M ∂u/∂t + K(t,u) = 0 -> solve for u
+EX-RK:  A(t,u,ki) = M ki    + K(ti,u0 + dt ∑_{j<i} a_ij * kj) = 0 -> solve for ki
+                  = M ki    + K(ti,ui) = 0
+
+For forward euler,          i = 1     -> ui = u0
+For other explicit methods, i = 1,…,s -> ui = u0 + dt ∑_{j<i} a_ij * kj
+
+For EX-RK, the Jacobian is always M. At each solve_step, compute and store M in
+nlop
```
Comment (santiagobadia) [inline src/ODEs/ODETools/EXRungeKutta.jl:99]: M is not a nlop, change notation

File: src/ODEs/ODETools/EXRungeKutta.jl [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+  M::AbstractMatrix
+end
+
+
+"""
+ODE:    A(t,u,∂u  = M ∂u/∂t + K(t,u) = 0 -> solve for u
+EX-RK:  A(t,u,ki) = M ki    + K(ti,u0 + dt ∑_{j<i} a_ij * kj) = 0 -> solve for ki
+                  = M ki    + K(ti,ui) = 0
+
+For forward euler,          i = 1     -> ui = u0
```
Comment (santiagobadia) [inline src/ODEs/ODETools/EXRungeKutta.jl:95]: don't you think this is just a particular case of the general case in which there is only one stage, I think you can accommodate it in the same implementation / def'on

**Review by santiagobadia — CHANGES_REQUESTED** [review]

File: my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl (hunk 1 above)
Comment (tamaratambyah) [reply my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl:1]: started clean branch of master fork. my_tests folder is removed

File: src/ODEs/ODETools/RungeKutta.jl (hunk 13 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/RungeKutta.jl:162]: clean version does not edit RungeKutta.jl

File: src/ODEs/ODETools/RungeKutta.jl (hunk 12 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/RungeKutta.jl:228]: clean version does not edit RungeKutta.jl

File: src/ODEs/ODETools/RungeKutta.jl (hunk 11 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/RungeKutta.jl:218]: clean version does not edit RungeKutta.jl

File: src/ODEs/ODETools/RungeKutta.jl (hunk 10 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/RungeKutta.jl:217]: clean version does not edit RungeKutta.jl

File: src/ODEs/ODETools/ForwardEuler.jl (hunk 7 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/ForwardEuler.jl:56]: removed in clean version

File: src/ODEs/Exports.jl (hunk 5 above)
Comment (tamaratambyah) [reply src/ODEs/Exports.jl:15]: removed in clean version

File: my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl (hunk 2 above)
Comment (tamaratambyah) [reply my_tests/TransientFEsTests/BoundaryHeatEquationTests.jl:2]: started clean branch of master fork.

File: src/Algebra/NLSolvers.jl (hunk 4 above)
Comment (tamaratambyah) [reply src/Algebra/NLSolvers.jl:55]: removed in clean version

File: src/ODEs/ODETools/DIRungeKutta.jl (hunk 6 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/DIRungeKutta.jl:3]: removed in clean version - can implement DIRK on a different branch

File: src/ODEs/ODETools/EXRungeKutta.jl (hunk 17 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/EXRungeKutta.jl:99]: changed notation in clean version

File: src/ODEs/ODETools/EXRungeKutta.jl (hunk 18 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/EXRungeKutta.jl:95]: updated this comment to be more general in clean version

File: src/ODEs/ODETools/EXRungeKutta.jl (hunk 15 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/EXRungeKutta.jl:44]: changed to EXRungeKuttaStageOperator < RungeKuttaNonlinearOperator.

File: src/ODEs/ODETools/EXRungeKutta.jl (hunk 16 above)
Comment (tamaratambyah) [reply src/ODEs/ODETools/EXRungeKutta.jl:52]: changed to l_cache

File: my_tests/dirk_step.jl (hunk 3 above)
Comment (tamaratambyah) [reply my_tests/dirk_step.jl:1]: removed in clean version

File: src/ODEs/TransientFETools/TransientFEOperators.jl (hunk 9 above)
Comment (tamaratambyah) [reply src/ODEs/TransientFETools/TransientFEOperators.jl:460]: There is no linear OperatorType in ODEOperators.jl. Of the OperatorTypes included, I think Nonlinear is the best choice.

**Comment by santiagobadia** [comment]
As discussed @tamaratambyah will create a new PR

**Pull request closed without merging**